packages/arb-validator/rollup/stakedNodeGraph.go

```diff
@@ -163,110 +168,140 @@ func (chain *StakedNodeGraph) generateNodePruneInfo(stakers *StakerSet) []pruneP
 	return prunesToDo
 }
 
-type confirmValidOpportunity struct {
-	nodeHash           common.Hash
-	deadlineTicks      common.TimeTicks
-	messages           []value.Value
-	logsAcc            common.Hash
-	vmProtoStateHash   common.Hash
-	stakerAddresses    []common.Address
-	stakerProofs       []common.Hash
-	stakerProofOffsets []*big.Int
-}
-
-type confirmInvalidOpportunity struct {
-	nodeHash           common.Hash
-	deadlineTicks      common.TimeTicks
-	challengeNodeData  common.Hash
-	branch             structures.ChildType
-	vmProtoStateHash   common.Hash
-	stakerAddresses    []common.Address
-	stakerProofs       []common.Hash
-	stakerProofOffsets []*big.Int
-}
-
 func (sng *StakedNodeGraph) generateNextConfProof(
 	currentTime common.TimeTicks,
-) (*confirmValidOpportunity, *confirmInvalidOpportunity) {
+) *valprotocol.ConfirmOpportunity {
 	stakerAddrs := make([]common.Address, 0)
 	sng.stakers.forall(func(st *Staker) {
 		stakerAddrs = append(stakerAddrs, st.address)
 	})
 	sort.Sort(SortableAddressList(stakerAddrs))
 
-	for _, successor := range sng.latestConfirmed.successorHashes {
-		if successor == zeroBytes32 {
-			continue
+	nodeOps := make([]valprotocol.ConfirmNodeOpportunity, 0)
+	conf := sng.latestConfirmed
+	keepChecking := true
+	confNodes := make([]*Node, 0)
+	for ; keepChecking; keepChecking = false {
+		for _, successor := range conf.successorHashes {
+			if successor == zeroBytes32 {
+				continue
+			}
+			node := sng.nodeFromHash[successor]
+
+			confirmable := sng.isConfirmableNode(
+				node,
+				currentTime,
+				stakerAddrs,
+			)
+			if confirmable {
+				var confOpp valprotocol.ConfirmNodeOpportunity
+				if node.linkType == valprotocol.ValidChildType {
+					if node.assertion == nil {
```

Comment:
Suggest a comment here explaining why we're not confirming the node because we don't know the assertion that made it.

And: Wasn't the assertion posted on-chain?  So how do we not know it?

Reply:
Added comment.
```go
// We need to know the contents of the actual assertion to confirm it
// We've only seen the hash accumulator of the messages before whereas this requires the full values
```